external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,517 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.storm.kafka.spout.KafkaSpoutConfig.FirstPollOffsetStrategy.EARLIEST;
+import static org.apache.storm.kafka.spout.KafkaSpoutConfig.FirstPollOffsetStrategy.LATEST;
+import static org.apache.storm.kafka.spout.KafkaSpoutConfig.FirstPollOffsetStrategy.UNCOMMITTED_EARLIEST;
+import static org.apache.storm.kafka.spout.KafkaSpoutConfig.FirstPollOffsetStrategy.UNCOMMITTED_LATEST;
+
+public class KafkaSpout<K, V> extends BaseRichSpout {
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaSpout.class);
+    private static final Comparator<KafkaSpoutMessageId> OFFSET_COMPARATOR = new OffsetComparator();
+
+    // Storm
+    protected SpoutOutputCollector collector;
+
+    // Kafka
+    private final KafkaSpoutConfig<K, V> kafkaSpoutConfig;
+    private transient KafkaConsumer<K, V> kafkaConsumer;
+    private transient boolean consumerAutoCommitMode;
+
+
+    // Bookkeeping
+    private KafkaSpoutStreams kafkaSpoutStreams;
+    private KafkaSpoutTupleBuilder<K, V> tupleBuilder;
+    private transient Timer commitTimer;                                    // timer == null for auto commit mode
+    private transient Timer logTimer;
+    private transient Map<TopicPartition, OffsetEntry> acked;         // emitted tuples that were successfully acked. These tuples will be committed periodically when the timer expires, on consumer rebalance, or on close/deactivate
+    private transient int maxRetries;                                 // Max number of times a tuple is retried
+    private transient boolean initialized;          // Flag indicating that the spout is still undergoing initialization process.
+    // Initialization is only complete after the first call to  KafkaSpoutConsumerRebalanceListener.onPartitionsAssigned()
+    private transient long numUncommittedOffsets;   // Number of offsets that have been polled and emitted but not yet been committed
+    private transient FirstPollOffsetStrategy firstPollOffsetStrategy;
+    private transient PollStrategy pollStrategy;
+
+
+    public KafkaSpout(KafkaSpoutConfig<K, V> kafkaSpoutConfig, KafkaSpoutTupleBuilder<K, V> tupleBuilder) {
+        this.kafkaSpoutConfig = kafkaSpoutConfig;                 // Pass in configuration
+        this.kafkaSpoutStreams = kafkaSpoutConfig.getKafkaSpoutStreams();
+        this.tupleBuilder = tupleBuilder;
+    }
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        initialized = false;
+
+        // Spout internals
+        this.collector = collector;
+        maxRetries = kafkaSpoutConfig.getMaxTupleRetries();
+        numUncommittedOffsets = 0;
```

Comment:
How is this different from topology.max.spout.pending?  If it were a maximum distance from commit instead of a count that would seem OK, but a count makes it seem like it is identical to topology.max.spout.pending.